src/wp-includes/media.php

```diff
@@ -793,6 +803,356 @@ function wp_get_attachment_image($attachment_id, $size = 'thumbnail', $icon = fa
+
+	// If sizes is passed as a string, just use the string.
+	if ( is_string( $args['sizes'] ) ) {
+		$size_list = $args['sizes'];
+
+	// Otherwise, breakdown the array and build a sizes string.
+	} elseif ( is_array( $args['sizes'] ) ) {
+
+		$size_list = '';
+
+		foreach ( $args['sizes'] as $size ) {
+
+			// Use 100vw as the size value unless something else is specified.
+			$size_value = ( $size['size_value'] ) ? $size['size_value'] : '100vw';
+
+			// If a media length is specified, build the media query.
+			if ( ! empty( $size['mq_value'] ) ) {
+
+				$media_length = $size['mq_value'];
+
+				// Use max-width as the media condition unless min-width is specified.
+				$media_condition = ( ! empty( $size['mq_name'] ) ) ? $size['mq_name'] : 'max-width';
+
+				// If a media_length was set, create the media query.
+				$media_query = '(' . $media_condition . ": " . $media_length . ') ';
+
+			} else {
+				// If no media length was set, `$media_query` is blank.
+				$media_query = '';
+			}
+
+			// Add to the source size list string.
+			$size_list .= $media_query . $size_value . ', ';
+		}
+
+		// Remove the trailing comma and space from the end of the string.
+		$size_list = substr( $size_list, 0, -2 );
+	}
+
+	// Return the `sizes` value as `$size_list` or false.
+	return ( $size_list ) ? $size_list : false;
+}
+
+/**
+ * Filters images in post content to add 'srcset' and 'sizes'.
+ *
+ * @since 4.4.0
+ *
+ * @param string $content The raw post content to be filtered.
+ * @return string Converted content with 'srcset' and 'sizes' added to images.
+ */
+function wp_resp_img( $content ) {
+
+	// Only match images in our uploads directory.
+	$uploads_dir = wp_upload_dir();
+	$path_to_upload_dir = $uploads_dir['baseurl'];
+
+	// Pattern for matching all images with a `src` from the uploads directory.
+	$pattern = '|<img ([^>]+' . preg_quote( $path_to_upload_dir ) . '[^>]+)>|i';
+	preg_match_all( $pattern, $content, $matches );
+
+	$images = $matches[0];
+	$ids = array();
+
+	foreach( $images as $image ) {
+		if ( preg_match( '/wp-image-([0-9]+)/i', $image, $class_id ) ) {
+			(int) $id = $class_id[1];
+			if ( $id ) {
+				$ids[] = $id;
+			}
+		}
+	}
+
+	if ( 0 < count( $ids ) ) {
+		/*
+		 * Warm object caches for use with wp_get_attachment_metadata.
+		 *
+		 * To avoid making a database call for each image, a single query
+		 * warms the object cache with the meta information for all images.
+		 **/
+		_prime_post_caches( $ids, false, true );
+	}
+
+	foreach( $matches[0] as $k => $image ) {
+		$match = array( $image, $matches[1][$k] );
+		$needle = $image;
+		$replacement = _wp_resp_img( $match );
+		if ( false === $replacement ) {
+			continue;
+		}
+		$content = str_replace( $image, $replacement, $content );
+	}
+
+	return $content;
+}
+
+/**
+ * Private filter callback used in 'wp_resp_img()'
+ *
+ * @access private
+ * @since 4.4.0
+ */
+function _wp_resp_img( $image ) {
```

Comment:
This was originally a preg_match callback function. We could probably merge these now.